presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionDescriptor.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.function;
+
+import com.facebook.presto.common.Subfield;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static com.facebook.presto.common.Subfield.allSubscripts;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableSet;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents the description of the function and contains the properties important for subfield extraction.
+ */
+public class FunctionDescriptor
+{
+    /**
+     * Indicates whether the function internally accessing subfields. This does not include the subfield access through the functional parameter (lambda) passed as an
+     * argument. If function accepts the functional argument that can access the subfields, then the <code>lambdaArgumentIndex</code> must contain the index of such
+     * functional argument and <code>lambdaArgumentToCallArgumentIndexMap</code> must contain the mapping between the lambda argument to the index of the argument of the
+     * CallExpression that this argument relates to.
+     */
+    private final boolean isAccessingSubfieldsInternally;
```

Comment:
Call this something like `isAccessingInputValues`.  This should return true anytime a function access the values of the map or array argument it gets.  No need to think about subfields specifically.

Reply:
I will rename it as you suggest but keep in mind that function can access the input values without accessing the subfields. For instance, it can check whether the values is `null` or not. The same with the `cardinality` function that accesses the array but only to get its length.

Reply:
We can ignore these subtleties and just say that in those corner cases `isAccessingInputValues=false`

Reply:
Yes, maybe there's a clearer way to describe the elements of an array or entries in a map, but i meant for the field name to convey whether it accesses any of those things.

Again the reasoning for this is because someone writing a function or anyone who hasn't been thinking about this a long time is going to have a hard time reasoning about subfields.  We want the descriptor to convey more generically  what qualities of the function we're looking to describe.  At a higher level, what would cause something to potentially access subfields internally? The answer is that if it accesses any of the elements of the array/map then it could access subfields.

Reply:
renamed to `isAccessingInputValues`